workflow/controller/workflowpod.go

```diff
@@ -282,8 +282,8 @@ func (woc *wfOperationCtx) createWorkflowPod(ctx context.Context, nodeName strin
 	pod.ObjectMeta.Annotations[common.AnnotationKeyDefaultContainer] = defaultContainer
 
 	// Add init container only if it needs input artifacts. This is also true for
-	// script templates (which needs to populate the script)
-	if len(tmpl.Inputs.Artifacts) > 0 || tmpl.GetType() == wfv1.TemplateTypeScript || woc.getContainerRuntimeExecutor() == common.ContainerRuntimeExecutorEmissary {
+	// resource and script templates (which needs to populate the script)
+	if len(tmpl.Inputs.Artifacts) > 0 || tmpl.GetType() == wfv1.TemplateTypeScript || tmpl.GetType() == wfv1.TemplateTypeResource || woc.getContainerRuntimeExecutor() == common.ContainerRuntimeExecutorEmissary {
```

Comment:
I think we should probably remove this check completely and always run the init container. But not in the PR. Tech debt.